imsApp/src/imsRecord.cc

```diff
@@ -3589,6 +3647,142 @@ static void post_fields( imsRecord *prec, unsigned short alarm_mask,
     return;
 }
 
+/******************************************************************************/
+static void io_controller ( struct ims_info *mInfo )
+{
+    char *s, *t, buf[2048];
+    const double wtimeout = 1.0;
+    const double rtimeout = 0.1;
+    asynStatus status;
+    int i, eomReason, wild;
+    size_t len;
+
+    for (;;) {
+	/* Check for new work */
+	mInfo->rMutex->lock();
+	for (i = 0; i < REQ_CNT; i++) {
+	    struct ims_asyn_req *req = &mInfo->req[i];
+
+	    if (req->state == IAR_SENDONLY) {
+		pasynOctetSyncIO->write( mInfo->pasynUser, req->out, 
+					 req->outsize, wtimeout, &len );
+		free_request_buf(mInfo, req);
+		continue;
+	    }
+	    if (req->state == IAR_SEND) {
+		pasynOctetSyncIO->write( mInfo->pasynUser, req->out, 
+					 req->outsize, wtimeout, &len );
+		req->state = IAR_RESPWAIT;
+		continue;
+	    }
+	}
+	mInfo->rMutex->unlock();
+
+	/* Try to read a message */
+	status = pasynOctetSyncIO->read( mInfo->pasynUser, buf, sizeof(buf) - 1,
+					 rtimeout, &len, &eomReason );
+	if (status != asynSuccess || len <= 0)
+	    continue;
+
+	if (len < sizeof(buf))
+	    buf[len] = 0;
+
+#if 0
+	/* Debug print. */
+	printf("< ");
+	for (size_t ii = 0; ii < len; ii++) {
+	    if (buf[ii] == '\r')
+		printf("\\r");
+	    else if (buf[ii] == '\n')
+		printf("\\n");
+	    else
+		putchar(buf[ii]);
+	}
+	putchar('\n');
+#endif
+
+	/* Is it even possible to have more than one message?!? */
+	for (s = buf, t = index(s, '\r'); s && *s; s = t) {
+	    t = index(s, '\r');
+	    if (t) {
```

Comment:
These two operations seem clever but dicey. For string parsing I really prefer the std::string implemention; there are often performance boosts as well not to mention the robusitification inherent in not dealing with raw pointers indexing into effectively arbitrary strings parsed of the line. 

(Again, if this PR is well tested and functional lets cut a future PR for the type of improvement I am suggesting in this comment. If this PR is still not _fully_ tested, it may be worth importing \<string\>)

Reply:
This has been removed.  See previous.